Simplenote/Controllers/PublishController.swift

```diff
@@ -0,0 +1,101 @@
+import Foundation
+
+@objc
+class PublishController: NSObject {
+    private var callbackMap: PublishMap
+    private let timerFactory: TimerFactory
+
+    init(timerFactory: TimerFactory = TimerFactory(), callbackMap: PublishMap = PublishMap()) {
+        self.timerFactory = timerFactory
+        self.callbackMap = callbackMap
+    }
+
+    func updatePublishState(for note: Note, to published: Bool, completion: @escaping (PublishState) -> Void) {
+        if note.published == published {
+            return
+        }
+
+        let wrapper = PublishListenWrapper(note: note, block: completion)
+        callbackMap.container[note.simperiumKey] = wrapper
+
+        SPTracker.trackEditorNotePublishEnabled(published)
+        changePublishState(for: note, to: published)
+
+        published ? update(wrapper, to: .publishing) : update(wrapper, to: .unpublishing)
+    }
+
+    @objc(didReceiveUpdateFromSimperiumForKey:)
+    func didReceiveUpdateFromSimperium(for key: NSString) {
+        guard let wrapper = callbackMap.container[key as String] else {
+            return
+        }
+
+        if wrapper.note.published && wrapper.note.publishURL != nil {
+            update(wrapper, to: .published)
+            return
+        }
+
+        if !wrapper.note.published {
+            update(wrapper, to: .unpublished)
+
+            return
+        }
+    }
+
+    private func update(_ wrapper: PublishListenWrapper, to state: PublishState) {
+        wrapper.block(state)
+        startTimer(in: wrapper)
+    }
+
+    private func changePublishState(for note: Note, to published: Bool) {
+        note.published = published
+        note.modificationDate = Date()
+        SPAppDelegate.shared().save()
+    }
+
+    fileprivate func removeListenerCallback(for key: String) {
+        callbackMap.container.removeValue(forKey: key)
+    }
+
+    fileprivate func startTimer(in wrapper: PublishListenWrapper) {
+        wrapper.timer.invalidate()
+
+        guard let key = wrapper.note.simperiumKey else {
+            return
+        }
+        wrapper.timer = timerFactory.scheduledTimer(with: Constants.timeOut, completion: {
+            self.removeListenerCallback(for: key)
+        })
+    }
+}
+
+enum PublishState {
+    case publishing
+    case published
+    case unpublishing
+    case unpublished
+}
+
+class PublishListenWrapper: NSObject {
+    let note: Note
+    let block: (PublishState) -> Void
+    var timer: Timer
```

Comment:
Rather than maintaining a Timer per Published Note, how about iffff:

A. We keep a (single) timer in the PublishController, which gets scheduled / rescheduled after you publish a Note. (YES scheduling two notes in a row will result in a single timer).

B. We just maintain a Timestamp in the Wrapper, and check if the listener has expired

```
expiration.timeIntervalSinceNow < Constants.timeout
```

Reasoning behind this change is: a structure such as this Wrapper should not contain this timer. The `Callback` runs `self.removeListenerCallback` from **PublishController** (and doesn't really call `PublishListenWrapper`.

*PLUS* this change would allow you to switch this to a `struct`, and drop the initializer!